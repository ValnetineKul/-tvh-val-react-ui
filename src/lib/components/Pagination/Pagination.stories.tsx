import { action } from '@storybook/addon-actions';
import type { ComponentProps } from 'react';
import React, { useState } from 'react';
import type { StoryTemplate } from '../../story-utils';
import { createMeta } from '../../story-utils';
import Pagination from './Pagination';

type Props = ComponentProps<typeof Pagination>;

type Args = Omit<Props, 'page' | 'numberOfPages' | 'onPageChange'>;

const Template = ((args: Args) => {
  const [page, setPage] = useState<number>(1);

  const handlePageChange = (v: number) => {
    setPage(v);
    action('onPageChange')(v);
  };

  return <Pagination {...args} page={page} numberOfPages={10} onPageChange={handlePageChange} />;
}) as StoryTemplate<Args>;

const WithItemsPerPageTemplate = ((args: Args) => {
  const numberOfItems = 100;
  const [page, setPage] = useState<number>(1);
  const [currentPageSizeValue, setCurrentPageSizeValue] = useState(10);
  const [numberOfPages, setNumberOfPages] = useState(numberOfItems / currentPageSizeValue);

  const handlePageChange = (v: number) => {
    setPage(v);
    action('onPageChange')(v);
  };

  const handleItemsPerPageChange = (v: number) => {
    setPage(1);
    setCurrentPageSizeValue(v);
    setNumberOfPages(numberOfItems / v);
    action('onItemsPerPageChange')(v);
  };

  return (
    <Pagination
      {...args}
      page={page}
      numberOfPages={numberOfPages}
      itemsPerPage={currentPageSizeValue}
      itemsPerPageLabel="Items per page:"
      onPageChange={handlePageChange}
      onItemsPerPageChange={handleItemsPerPageChange}
    />
  );
}) as StoryTemplate<Args>;

const InfinitePages = ((args: Args) => {
  const [page, setPage] = useState<number>(1);
  const [maxPages, setMaxPages] = useState<number>(page);

  const handlePageChange = (v: number) => {
    setPage(v);
    setMaxPages((m) => Math.max(m, v));
    action('onPageChange')(v);
  };

  return (
    <Pagination
      {...args}
      page={page}
      numberOfPages={maxPages < 10 ? maxPages + 1 : 10}
      isNumberOfPagesUnknown={maxPages < 10}
      onPageChange={handlePageChange}
    />
  );
}) as StoryTemplate<Args>;

export const Primitive = Template.bind({});
Primitive.args = { disabled: false };
Primitive.storyName = 'pagination';

export const WithItemsPerPage = WithItemsPerPageTemplate.bind({});
WithItemsPerPage.args = { disabled: false };
WithItemsPerPage.storyName = 'Items per page';

export const WithInfinitePages = InfinitePages.bind({});
WithInfinitePages.args = { disabled: false };
WithInfinitePages.storyName = 'Infinite pages';

export default createMeta({
  component: Pagination,
  title: 'Components/Paginations/Pagination',
  argTypes: {
    page: {
      control: {
        disable: true,
      },
      table: {
        type: {
          detail: 'The current page',
        },
      },
    },
    numberOfPages: {
      control: {
        disable: true,
      },
      table: {
        type: {
          detail: 'The total number of pages',
        },
      },
    },
    itemsPerPage: {
      control: {
        disable: true,
      },
      table: {
        type: {
          detail: 'The total number of items per page',
        },
      },
    },
    itemsPerPageLabel: {
      control: {
        disable: true,
      },
      table: {
        type: {
          summary: 'Label for Items per page',
        },
      },
    },
    itemsPerPageOptions: {
      control: {
        disable: true,
      },
      table: {
        type: {
          detail: 'Options: { value: 10 } or { label: "View all", value: "viewAll" }.',
        },
      },
    },
  },
});
