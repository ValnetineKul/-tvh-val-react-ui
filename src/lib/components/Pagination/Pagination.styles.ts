import { makeStyles } from '../../../themes/core';

const useStyles = makeStyles({ name: 'Pagination' })((theme) => ({
  root: {
    display: 'flex',
    flexFlow: 'column-reverse',
    alignItems: 'center',
    [theme.breakpoints.up('sm')]: {
      flexFlow: 'row',
    },
  },
  list: {
    '& li': {
      display: 'flex',
    },
    '& li > div': {
      marginRight: theme.layoutSpacing['spacing/300'],
    },
    '& li:last-child > div': {
      marginRight: theme.layoutSpacing['spacing/000'],
    },
  },
  itemsPerPage: {
    marginTop: theme.layoutSpacing['spacing/300'],
    [theme.breakpoints.up('sm')]: {
      marginTop: 0,
      marginRight: 'auto',
    },
  },
}));

export default useStyles;
