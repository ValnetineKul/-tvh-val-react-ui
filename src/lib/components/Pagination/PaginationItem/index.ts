import PaginationItem from './PaginationItem';

export type { PaginationItemProps } from './PaginationItem';
export default PaginationItem;
