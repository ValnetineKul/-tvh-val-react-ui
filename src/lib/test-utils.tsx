import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
