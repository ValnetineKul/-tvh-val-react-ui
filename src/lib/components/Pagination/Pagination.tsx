import type { PaginationRenderItemParams } from '@mui/material/Pagination';
import { default as MuiPagination } from '@mui/material/Pagination';
import type { FC } from 'react';
import React from 'react';
import ListOptionSelector from '../ListOptionSelector';
import useStyles from './Pagination.styles';
import PaginationItem from './PaginationItem';
import type { DataAttributes } from '../../types/common';
import { useScreenSize } from '../../hooks';

export interface PaginationProps {
  page: number;
  numberOfPages: number;
  isNumberOfPagesUnknown?: boolean;
  itemsPerPage?: number | string;
  itemsPerPageLabel?: string;
  itemsPerPageOptions?: { label?: string; value: number | string }[];
  disabled?: boolean;
  onPageChange: (page: number, e?: React.ChangeEvent<unknown>) => void;
  onItemsPerPageChange?: (itemsPerPage: number | string) => void;
  perPageSelectorProps?: React.HTMLAttributes<HTMLButtonElement> & DataAttributes;
  pageButtonProps?: React.HTMLAttributes<HTMLButtonElement> & DataAttributes;
  pageNextProps?: React.HTMLAttributes<HTMLButtonElement> & DataAttributes;
  pagePrevProps?: React.HTMLAttributes<HTMLButtonElement> & DataAttributes;
}

const visiblePaginationItemsOnDesktop = 7;
const visiblePageItemsOnDesktop = 4;
const visiblePageItemsOnMobile = 3;

let shouldDeletePageAfterEllipsis = false;

const getPagePropsByType = (
  type: PaginationRenderItemParams['type'],
  props: Partial<Record<string, React.HTMLAttributes<HTMLButtonElement> & DataAttributes>>
) => {
  if (type === 'next') {
    return props.pageNextProps;
  }
  if (type === 'previous') {
    return props.pagePrevProps;
  }
  return props.pageButtonProps;
};

const Pagination: FC<PaginationProps> = ({
  page,
  numberOfPages,
  isNumberOfPagesUnknown,
  itemsPerPage,
  itemsPerPageLabel = '',
  onPageChange,
  onItemsPerPageChange,
  disabled = false,
  itemsPerPageOptions = [{ value: 10 }, { value: 25 }, { value: 50 }, { value: 100 }],
  perPageSelectorProps,
  pageButtonProps,
  pageNextProps,
  pagePrevProps,
}) => {
  const { classes } = useStyles();

  const { isTabletUp } = useScreenSize('Tablet');
  const siblingCount = isTabletUp ? 1 : 0;

  return (
    <div className={classes.root}>
      {itemsPerPage && onItemsPerPageChange ? (
        <ListOptionSelector
          label={itemsPerPageLabel}
          value={itemsPerPage}
          options={itemsPerPageOptions}
          className={classes.itemsPerPage}
          disabled={disabled}
          onChange={(v: number | string) => onItemsPerPageChange(v)}
          buttonProps={perPageSelectorProps}
        />
      ) : null}
      <MuiPagination
        count={numberOfPages}
        siblingCount={siblingCount}
        page={page}
        disabled={disabled}
        onChange={(e, newPage) => onPageChange(newPage, e)}
        classes={{ ul: classes.list }}
        renderItem={({ type, page: pageNumber, selected, disabled: isDisabled, onClick, ...prop }) => {
          if (type === 'first' || type === 'last') return null;
          const isLastPageElement = pageNumber === numberOfPages && type === 'page';
          const paginationItemsAhead = isTabletUp ? visiblePageItemsOnDesktop : visiblePageItemsOnMobile;
          const buttonProps =
            getPagePropsByType(type, {
              pageNextProps,
              pagePrevProps,
              pageButtonProps,
            }) || {};

          const shouldChangeLastItem = isNumberOfPagesUnknown && isLastPageElement;
          const shouldDeleteLastItem =
            shouldChangeLastItem &&
            numberOfPages - page >= paginationItemsAhead &&
            numberOfPages > visiblePaginationItemsOnDesktop;

          if (
            isNumberOfPagesUnknown &&
            isTabletUp &&
            pageNumber - page >= 2 &&
            numberOfPages >= visiblePaginationItemsOnDesktop
          ) {
            shouldDeletePageAfterEllipsis = true;
          }

          if (
            isNumberOfPagesUnknown &&
            isTabletUp &&
            numberOfPages === visiblePaginationItemsOnDesktop &&
            type === 'page' &&
            page > visiblePageItemsOnDesktop
          ) {
            if (pageNumber === 2) {
              return null;
            }

            if (pageNumber === 3) {
              return (
                <PaginationItem
                  type="start-ellipsis"
                  onClick={() => {}}
                  page={3}
                  selected={false}
                  disabled={isDisabled}
                  {...buttonProps}
                />
              );
            }
          }

          if (
            isNumberOfPagesUnknown &&
            isTabletUp &&
            numberOfPages - page === 1 &&
            numberOfPages >= visiblePaginationItemsOnDesktop &&
            pageNumber + 4 === numberOfPages
          ) {
            return null;
          }

          const PaginationItemEl = (
            <PaginationItem
              type={type}
              page={pageNumber}
              selected={selected}
              disabled={isDisabled}
              onClick={onClick}
              aria-label={(prop as unknown as Record<string, string>)['aria-label']}
              aria-current={(prop as unknown as Record<string, boolean>)['aria-current']}
              {...buttonProps}
            />
          );

          if (shouldChangeLastItem) {
            const paginationEl = !shouldDeletePageAfterEllipsis ? PaginationItemEl : null;
            if (shouldDeletePageAfterEllipsis) {
              shouldDeletePageAfterEllipsis = false;
            }
            return shouldDeleteLastItem ? null : (
              <>
                {paginationEl}
                <PaginationItem
                  type="end-ellipsis"
                  onClick={() => {}}
                  page={numberOfPages}
                  selected={false}
                  disabled={isDisabled}
                  {...buttonProps}
                />
              </>
            );
          }
          return PaginationItemEl;
        }}
      />
    </div>
  );
};

export default Pagination;
