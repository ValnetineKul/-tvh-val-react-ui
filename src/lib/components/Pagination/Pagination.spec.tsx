import React from 'react';
import { render, screen, userEvent, within, initResponsiveTest } from '../../test-utils';
import Pagination from './Pagination';

jest.mock('@mui/material/useMediaQuery', () => ({
  __esModule: true,
  default: jest.fn().mockReturnValue(true),
}));

describe('Pagination', () => {
  it('Should render correctly', () => {
    const handlePageChange = jest.fn();
    const handleItemsPerPageChange = jest.fn();
    const { container } = render(
      <Pagination
        page={2}
        numberOfPages={3}
        itemsPerPage={10}
        onPageChange={handlePageChange}
        onItemsPerPageChange={handleItemsPerPageChange}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it.each([
    ['page 2', 2],
    ['Go to page 3', 3],
    ['Go to previous page', 1],
    ['Go to next page', 3],
  ])('Should emit the new page on clicking button %s', (buttonRoleName, expectedPage) => {
    const handlePageChange = jest.fn();
    render(<Pagination page={2} numberOfPages={3} onPageChange={handlePageChange} />);

    const button = screen.getByRole('button', { name: buttonRoleName });
    userEvent.click(button);
    expect(handlePageChange).toHaveBeenCalledWith(expectedPage, expect.anything());
  });

  it('Should emit the value on items per page change', () => {
    initResponsiveTest('Desktop');

    const handleItemsPerPageChange = jest.fn();
    const handlePageChange = jest.fn();
    render(
      <Pagination
        page={2}
        numberOfPages={3}
        itemsPerPage={10}
        onPageChange={handlePageChange}
        onItemsPerPageChange={handleItemsPerPageChange}
      />
    );

    const dropdown = screen.getByRole('button', { name: '10' });
    expect(dropdown.textContent).toMatch(/10/);
    userEvent.click(dropdown);

    expect(screen.getByRole('tooltip')).toBeInTheDocument();

    const fiftyButton = screen.getByRole('button', { name: '50' });
    userEvent.click(fiftyButton);

    expect(handleItemsPerPageChange).toHaveBeenCalledWith(50);
  });

  it.each([
    [2, 3, 2],
    [10, 20, 3],
    [7, 10, 2],
    [5, 10, 3],
    [2, 10, 3],
    [3, 10, 3],
    [5, 7, 2],
  ])('Should render ellipsis at the end when page are infinite', (page, numberOfPages, orderFromEnd) => {
    const handleItemsPerPageChange = jest.fn();
    const handlePageChange = jest.fn();

    render(
      <Pagination
        page={page}
        numberOfPages={numberOfPages}
        itemsPerPage={1}
        onPageChange={handlePageChange}
        onItemsPerPageChange={handleItemsPerPageChange}
        isNumberOfPagesUnknown
      />
    );

    const paginationItems = screen.getAllByRole('listitem');
    const infiniteEllipsis = within(paginationItems[paginationItems.length - orderFromEnd]).getByText(/\u2026/i);

    expect(infiniteEllipsis).toBeInTheDocument();
  });
});
