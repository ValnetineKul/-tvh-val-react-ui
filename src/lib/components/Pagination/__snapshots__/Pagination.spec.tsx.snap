// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Should render correctly 1`] = `
<div
  class="tss-pjh3kl-Pagination-root"
>
  <div
    class="tss-bd6lsj-ListOptionSelector-root-Pagination-itemsPerPage"
  >
    <span
      class="tss-5-Typography-weightRegular-ref tss-3ton9f-Typography-root-Typography-weightRegular-ListOptionSelector-label-Typography-body500"
    />
    <div
      class="tss-1kammvl-Surface-root-Button-surface"
    >
      <div
        class="tss-pyz8vm-SpinnerWithBackdrop-root"
      >
        <div
          class=""
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButtonBase-root tss-3-Button-variantLink-ref tss-3-Button-linkTypePrimary-ref tss-8544pl-Button-root-Button-variantLink-Button-linkTypePrimary-ListOptionSelector-value mui-10gi7f1-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            10
            <span
              class="tss-1gywngm-Button-endIcon MuiButton-endIcon MuiButton-iconSizeMedium mui-9dzbxy-MuiButton-endIcon"
            >
              <div
                class="tss-4fpoey-Icon-root-Icon-sizeSm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
                  color="inherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  angleDown.svg
                </svg>
              </div>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <nav
    aria-label="pagination navigation"
    class="MuiPagination-root MuiPagination-text mui-1oj2twp-MuiPagination-root"
  >
    <ul
      class="tss-1lziobi-Pagination-list MuiPagination-ul mui-wjh20t-MuiPagination-ul"
    >
      <li>
        <div>
          <button
            aria-label="Go to previous page"
            class="MuiButtonBase-root tss-1bnvnij-ButtonBase-root-ButtonBase-focusOut-IconButton-root-PaginationItem-arrowButton mui-10d1a0h-MuiButtonBase-root"
            tabindex="0"
            type="button"
          >
            <div
              class="tss-w1sjks-Icon-root-Icon-sizeMd"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
                color="inherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                angleLeft.svg
              </svg>
            </div>
          </button>
        </div>
      </li>
      <li>
        <div>
          <div
            class="tss-1kammvl-Surface-root-Button-surface"
          >
            <div
              class="tss-pyz8vm-SpinnerWithBackdrop-root"
            >
              <div
                class=""
              >
                <button
                  aria-label="Go to page 1"
                  class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButtonBase-root tss-3-Button-variantLink-ref tss-3-Button-linkTypeDefault-ref tss-1mri05j-Button-root-Button-variantLink-Button-linkTypeDefault-PaginationItem-page-PaginationItem-linkButton mui-10gi7f1-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  1
                </button>
              </div>
            </div>
          </div>
        </div>
      </li>
      <li>
        <div>
          <div
            class="tss-1bm80c1-Surface-root-Button-surface"
          >
            <div
              class="tss-pyz8vm-SpinnerWithBackdrop-root"
            >
              <div
                class=""
              >
                <button
                  aria-current="true"
                  aria-label="page 2"
                  class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButtonBase-root tss-6qgwg1-Button-root-Button-sizeMd-Button-button-PaginationItem-page mui-10gi7f1-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  2
                </button>
              </div>
            </div>
          </div>
        </div>
      </li>
      <li>
        <div>
          <div
            class="tss-1kammvl-Surface-root-Button-surface"
          >
            <div
              class="tss-pyz8vm-SpinnerWithBackdrop-root"
            >
              <div
                class=""
              >
                <button
                  aria-label="Go to page 3"
                  class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButtonBase-root tss-3-Button-variantLink-ref tss-3-Button-linkTypeDefault-ref tss-1mri05j-Button-root-Button-variantLink-Button-linkTypeDefault-PaginationItem-page-PaginationItem-linkButton mui-10gi7f1-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  3
                </button>
              </div>
            </div>
          </div>
        </div>
      </li>
      <li>
        <div>
          <button
            aria-label="Go to next page"
            class="MuiButtonBase-root tss-1bnvnij-ButtonBase-root-ButtonBase-focusOut-IconButton-root-PaginationItem-arrowButton mui-10d1a0h-MuiButtonBase-root"
            tabindex="0"
            type="button"
          >
            <div
              class="tss-w1sjks-Icon-root-Icon-sizeMd"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
                color="inherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                angleRight.svg
              </svg>
            </div>
          </button>
        </div>
      </li>
    </ul>
  </nav>
</div>
`;
